### PR TITLE
bugfix: read the request body from the temporary file if it was cached.

### LINKED ISSUE #1784: bug: /admin/routes return {"error_msg":"missing configurations"}
### Issue description
If there are too many uris, return {"error_msg":"missing configurations"}

cURL：
curl -v -s ${GATEWAY}/apisix/admin/routes -H "X-API-KEY: $ADMINKEY"  -X POST -d "
    {
        \"uris\": [$uris],
        \"hosts\": [$DOMAINS],
        \"service_id\": \"1\",
        \"desc\": \"apisix\"
    }"

error.log：
 2020/06/30 16:36:06 [warn] 119916#119916: *857380 a client request body is buffered to a temporary file /usr/local/apisix/client_body_temp/0000000003, client: 1.14.31.17, server: , request: "POST /apisix/admin/routes HTTP/1.1", host: "1.2.6.2:9080"

### Environment

* apisix version (cmd: `apisix version`):1.3
* OS:CentOS

### PR DESCRIPTION
**Pull request by membphis**
### What this PR does / why we need it:

as title, fixed #1784

perf: update the jsonschema to v0.9 .

used hash object to check if there was a duplicated value in the array object,
the current way is 10~100 times than the old way. 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: rockspec/apisix-master-0.rockspec [hunk 1]
```diff
@@ -47,7 +47,7 @@ dependencies = {
     "luafilesystem = 1.7.0-2",
     "lua-tinyyaml = 0.1",
     "lua-resty-prometheus = 1.1",
-    "jsonschema = 0.8",
+    "jsonschema = 0.9",
```
Comment (moonming) [inline rockspec/apisix-master-0.rockspec:50]: should not modify this line

File: t/admin/health-check.t [hunk 2]
```diff
@@ -288,7 +288,7 @@ GET /t
 GET /t
 --- error_code: 400
 --- response_body
-{"error_msg":"invalid configuration: property \"upstream\" validation failed: property \"checks\" validation failed: property \"active\" validation failed: property \"healthy\" validation failed: property \"http_statuses\" validation failed: expected unique items but items 2 and 1 are equal"}
+{"error_msg":"invalid configuration: property \"upstream\" validation failed: property \"checks\" validation failed: property \"active\" validation failed: property \"healthy\" validation failed: property \"http_statuses\" validation failed: expected unique items but items 1 and 2 are equal"}
```
Comment (moonming) [inline t/admin/health-check.t:291]: Why did you change this test case?

**Comment by membphis** [comment]
https://github.com/apache/incubator-apisix/pull/1863/files#diff-00625723b6e737f3cdb18af67165b70fR1899-R1901

@moonming in this test case, we found two bugs.

1. big request body for admin API
2. slower to detect whether the objects in the test array are unique (jsonschema).

image

**Comment by moonming** [comment]
one PR only do one think.

Thanks,
Ming Wen
Twitter: _WenMing


YuanSheng Wang <notifications@github.com> 于2020年7月20日周一 下午1:37写道：

File: apisix/admin/init.lua [hunk 3]
```diff
@@ -117,10 +117,13 @@ local function run()
         core.response.exit(404)
     end
 
-    ngx.req.read_body()
-    local req_body = ngx.req.get_body_data()
-
+    local req_body = core.request.get_body()
     if req_body then
+        if #req_body > 1024 * 1024 * 1.5 then
```
Comment (spacewander) [inline apisix/admin/init.lua:122]: We need to detect the size (use `stat`, for example) before reading it. It doesn't make sense to do the check if we have already read the very BIG file.

File: apisix/admin/init.lua [hunk 4]
```diff
@@ -117,10 +117,13 @@ local function run()
...
+            core.log.error("maximum request body is 1.5mb, but got ", #req_body)
```
Comment (spacewander) [inline apisix/admin/init.lua:123]: IMHO, we can use MiB to indicate the size is 1024 base.

File: apisix/core/request.lua [hunk 5]
```diff
@@ -94,7 +99,36 @@ function _M.get_remote_client_port(ctx)
         ctx = ngx.ctx.api_ctx
     end
     return tonumber(ctx.var.remote_port)
-  end
+end
+
+
+local function get_file(file_name)
+    local f = io_open(file_name, 'r')
```
Comment (spacewander) [inline apisix/core/request.lua:106]: Better to check the second error returned value before using the `f`.

File: apisix/admin/init.lua (hunk 3 above)
Comment (moonming) [reply apisix/admin/init.lua:122]: agreed.
